### ISSUE TITLE
BUG: Incorrect assignment to recarray items

### ISSUE THREAD

**Issue by magistere** [numpy/numpy#3561]
Assigning to field of recarray is not working correctly. Example script is below.

```python
import numpy as np

rec = np.recarray(1, dtype=[('x', float, 5)])

# Bug behaviour
rec[0].x = 1
print rec[0] # should print: (array([ 1.,  1.,  1.,  1.,  1.]),)
             # but prints uninitialized float values

# Correct behaviour
rec.x = 2
print rec[0] # prints: (array([ 2.,  2.,  2.,  2.,  2.]),)
```

**Comment by ahaldane** [comment]
This is a bug in ndarray, nor recarray. For example:

    >>> a = np.zeros(1, dtype=[('x', float, 5)])
    >>> a[0]
    ([0.0, 0.0, 0.0, 0.0, 0.0],)
    >>> a[0]['x']
    array([ 0.,  0.,  0.,  0.,  0.])
    >>> a[0]['x'] = 1
    >>> a
    array([([1.0, 0.0, 0.0, 0.0, 0.0],)], 
          dtype=[('x', '<f8', (5,))])

In particular, I think it is a bug in ndarray.setfield. 

I think ndarray.getfield and ndarray.setfield are buggy (eg, see the segfauly I cause in https://github.com/numpy/numpy/issues/3256#issuecomment-71697791). I think it might be possible to make them simply call np.view in the right way to solve all these problems. But it will wait until some other issues with views are solved (see https://github.com/numpy/numpy/pull/5508 https://github.com/numpy/numpy/issues/3256 https://github.com/numpy/numpy/issues/3286 https://github.com/numpy/numpy/issues/2599)

**Comment by mbyt** [comment]
@ahaldane, I have proposed a possible solution to this (raise at least an error) in the pull request #4556.

Does this really have to wait? It's just a 15 line long change / patch in production code.

**Comment by ahaldane** [comment]
Oh, I missed that link to your pull request above.

I took at your pull request and it seems like a good fix for now. I tested it and it worked, and the code looks OK (though I am a numpy code novice).

What I mentioned about waiting is that I think this problem (plus others) might be solved by making getfield and setfield go through the 'view' machinery. This way all the checks that happen for views would automatically happen for get/setfield.

I imagine setfield could do something like:

    >>> t,o = x[0].dtype.fields['field']
    >>> xv = x[0].view(dtype({'names': ['field'], 
                              'formats': [t], 
                              'offsets': [o], 
                              'itemsize': x.dtype.itemsize}))
    >>> #then do array_assign(xv, 4*ones)

Taking a view would also solve the following problem:

    >>> x[0].setfield(1, 'i8', x.dtype.fields['b'][1])
    >>> x
    zsh: segmentation fault (core dumped)

Now I admit I have not looked deeply into the assignment machinery yet so maybe this isn't possible...
